src/dxtbx/model/beam.py

```diff
@@ -57,61 +72,77 @@ class BeamFactory:
-    def from_phil(params, reference=None):
+    def from_phil(
+        params: libtbx.phil.scope_extract,
+        reference: Union[Beam, PolychromaticBeam] = None,
+    ) -> Union[Beam, PolychromaticBeam]:
         """
         Convert the phil parameters into a beam model
-
         """
+
         # Check the input
         if reference is None:
-            beam = Beam()
+            beam = (
+                PolychromaticBeam() if params.beam.type == "polychromatic" else Beam()
+            )
         else:
             beam = reference
 
         # Set the parameters
-        if params.beam.wavelength is not None:
-            beam.set_wavelength(params.beam.wavelength)
-        elif reference is None:
-            raise RuntimeError("No wavelength set")
+        if params.beam.type == "monochromatic":
+            if params.beam.wavelength is not None:
+                beam.set_wavelength(params.beam.wavelength)
+            elif reference is None:
+                raise RuntimeError("No wavelength set")
         if params.beam.direction is not None:
             beam.set_direction(params.beam.direction)
         elif reference is None:
             raise RuntimeError("No beam direction set")
+
+        if params.beam.divergence is not None:
+            beam.set_divergence(params.beam.divergence)
+        if params.beam.sigma_divergence is not None:
+            beam.set_sigma_divergence(params.beam.sigma_divergence)
         if params.beam.polarization_normal is not None:
             beam.set_polarization_normal(params.beam.polarization_normal)
         if params.beam.polarization_fraction is not None:
             beam.set_polarization_fraction(params.beam.polarization_fraction)
+        if params.beam.transmission is not None:
+            beam.set_transmission(params.beam.transmission)
+        if params.beam.flux is not None:
+            beam.set_flux(params.beam.flux)
 
         return beam
 
     @staticmethod
-    def from_dict(d, t=None):
-        """Convert the dictionary to a beam model
+    def from_dict(dict: Dict, template: Dict = None) -> Union[Beam, PolychromaticBeam]:
+        """Convert the dictionary to a beam model"""
 
-        Params:
-            d The dictionary of parameters
-            t The template dictionary to use
+        if template is not None:
+            if "__id__" in dict and "__id__" in template:
+                assert (
+                    dict["__id__"] == template["__id__"]
+                ), "Beam and template dictionaries are not the same type."
 
-        Returns:
-            The beam model
-        """
-        if d is None and t is None:
+        if dict is None and template is None:
             return None
-        joint = t.copy() if t else {}
-        joint.update(d)
+        joint = template.copy() if template else {}
+        joint.update(dict)
 
         # Create the model from the joint dictionary
+        if "__id__" in joint and joint["__id__"] == "polychromatic":
```

Comment:
`joint.get("__id__", "") == "polychromatic"`? Would be tidier (and I think is the intent of `dict.get()`

Reply:
I think you can just make that suggestion `joint.get("__id__") == "polychromatic"` (default is `None`)

Reply:
Going to leave it as Graeme's suggestion for the sake of the polar bears.